### PR TITLE
ci(static): Add static code tests

### PR DESCRIPTION
**Pull request by merkata**
This PR implements bandit as a static code analysis tool.

The scan includes src/ and tests/ directories, scanned recursively. Third party code (like venv/ and lib/) is not scanned currently.

The `assert_used` test (B101) is not enabled, it is noisy on test files and is also not a security or other type of concern.

The output is made part of the report for lint and unit tests.

Note I've preserved the workflow step to be `lint-and-unit-test` for now, suggestions are welcome on whether I should change it to reflect it includes static code analysis explicitly or it is fine as is.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -84,3 +84,11 @@ deps =
...
 commands =
     pytest -v --tb native --ignore={[vars]tst_path}unit --log-cli-level=INFO -s {posargs}
+
+[testenv:static]
+description = Run static analysis tests
+deps =
+    bandit
+    -r{toxinidir}/requirements.txt
+commands =
+    bandit -x {toxinidir}/venv -s B101 -r {[vars]src_path} {[vars]tst_path}
```
Comment (arturo-seijas) [inline tox.ini:94]: Can we add the exclusions in the toml file instead?

File: .github/files/extract_test_result.py [hunk 2]
```diff
@@ -20,12 +20,16 @@ def no_color(s):
     unit_result = result["testenvs"]["unit"]["test"]
     unit_success = unit_result[0]["retcode"] == 0
     unit_output = unit_result[0]["output"]
+    static_result = result["testenvs"]["static"]["test"]
```
Comment (arturo-seijas) [inline .github/files/extract_test_result.py:23]: Can you provide a sample comment to see how this looks like?

**Review by arturo-seijas — CHANGES_REQUESTED** [review]

File: tox.ini (hunk 1 above)
Comment (merkata) [reply tox.ini]: Added .bandit and I'm referencing it in the tox.ini.

File: .github/files/extract_test_result.py (hunk 2 above)
Comment (merkata) [reply .github/files/extract_test_result.py:23]: Sure, here is one (I'm still struggling on making javascript post it bellow the coverage report :/ ):
```
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: B101
[main]	INFO	running on Python 3.10.4
Run started:2022-09-20 10:42:16.416324

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 985
	Total lines skipped (#nosec): 0
	Total potential issues skipped due to specifically being disabled (e.g., #nosec BXXX): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
Files skipped (0):
```

**Comment by merkata** [comment]
This is how the report will look like:

Test coverage report for 234234
```
Name           Stmts   Miss Branch BrPart  Cover   Missing
----------------------------------------------------------
src/charm.py     220     22     66      6    88%   80->exit, 91, 95, 289->exit, 295->exit, 435-437, 468->exit, 483-505, 509-510
----------------------------------------------------------
TOTAL            220     22     66      6    88%
```
Static code analysis report
```
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: B101
[main]	INFO	running on Python 3.10.4
Run started:2022-09-20 10:42:16.416324

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 985
	Total lines skipped (#nosec): 0
	Total potential issues skipped due to specifically being disabled (e.g., #nosec BXXX): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
Files skipped (0):
```

Let me know whether something needs to be changed.

File: tox.ini (hunk 1 above)
Comment (merkata) [reply tox.ini]: Depends on what we want to do with the B101 `asserts_used` check - alter the check to skip test files or just don't run the check for any file we're scanning. Looking at the documentation, excluding directories and checks is possible in an INI file and if .bandit is in INI format then we don't even have to pass it in the tox.ini file as `-c <config>`, it should be picked up automatically. However, the same documentation mentions that if you want to alter behavior for a check, you can do it via a YAML file (and the example is in YAML as well).

Comment (arturo-seijas) [reply tox.ini]: I understand from your comment we don't need to pass the '-c' option with the current config files, right? Maybe we can just remove that option and keep the .bandit. Thoughts?

File: .github/files/comment_pr.js [hunk 3]
```diff
@@ -39,5 +40,5 @@ module.exports = async ({github, context}) => {
         const unitReport = report.reports.unit.output.trim();
         await createComment(`Unit tests failed for ${sha}\n\`\`\`\n${unitReport}\n\`\`\``);
     }
-    await createComment(`Test coverage report for ${sha}\n\`\`\`\n${coverageReport}\n\`\`\``);
+    await createComment(`Test coverage report for ${sha}\n\`\`\`\n${coverageReport}\n\`\`\`\nStatic code analysis report\n\`\`\`\n${staticReport}\n\`\`\``);
```
Comment (weiiwang01) [inline .github/files/comment_pr.js:43]: This line is a little bit too long, maybe change to something like this?

```javascript
    await createComment(
`Test coverage report for ${sha}
\`\`\`
${coverageReport}
\`\`\`
Static code analysis report
\`\`\`
${staticReport}
\`\`\``
);

```

File: tox.ini (hunk 1 above)
Comment (merkata) [reply tox.ini]: That's what the documentation says, that's how the reality behaves though (the default file with INI configuration does not work, even with taking the example config from the documentation):

```shell
merkata@mastodon:~/dev/merkata-indico-operator$ cat .bandit 
[bandit]
exclude = venv
skips = B101
merkata@mastodon:~/dev/merkata-indico-operator$ bandit -r src/ tests/
...some lines omitted...
>> Issue: [B101:assert_used] Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.
   Severity: Low   Confidence: High
   CWE: CWE-703 (https://cwe.mitre.org/data/definitions/703.html)
   Location: tests/integration/test_charm.py:61:12
   More Info: https://bandit.readthedocs.io/en/1.7.4/plugins/b101_assert_used.html
60	        else:
61	            assert stdout.count("0/3") == 2

--------------------------------------------------

Code scanned:
	Total lines of code: 985
	Total lines skipped (#nosec): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 6
...some lines omitted
```

I suggest we keep the .bandit with YAML and point directly to it. If you pass in `--ini .bandit` it works, but this disables assert rule globally for all file, whereas with the YAML it is skipped for specific files (that are test files in our case).
Since this effort is 5 story points, I/we could look at where and why this fails in the code of bandit.

File: .github/files/comment_pr.js (hunk 3 above)
Comment (merkata) [reply .github/files/comment_pr.js]: Looks nicer indeed, refactored. Thanks!

File: tox.ini (hunk 1 above)
Comment (arturo-seijas) [reply tox.ini]: I don't really think it'd pay off to investigate. It's good as it is

**Review by arturo-seijas — APPROVED** [review]
LGTM

**Review by mthaddon — APPROVED** [review]
Looks great, thanks

File: .bandit [hunk 4]
```diff
@@ -0,0 +1,4 @@
+assert_used:
```
Comment (jdkandersson) [inline .bandit:1]: I know the PR has closed, was just reading out of interest, this could be done in `pyproject.toml` I believe: https://bandit.readthedocs.io/en/latest/config.html

Comment (merkata) [reply .bandit:1]: Sounds good @jdkandersson , since we have a `pyproject.toml` already, we could add it here. Note I've had some issues with the config as described in the documentation and as it works in reality, specifically I found that a YAML config picks up the configuration values, while with TOML and INI I've had issues. I suggest opening an issue (I can open one as well, let me know) for bandit to be configured via pyproject.toml. Thanks for the comment and well spotted!

Comment (jdkandersson) [reply .bandit:1]: @merkata Yes please create the ticket

Comment (merkata) [reply .bandit:1]: https://github.com/canonical/indico-operator/issues/141
